### PR TITLE
Allow create service using udp

### LINKED ISSUE #47337: "kubectl create service" should allow UDP port
"kubectl create service" only supports TCP port. We should allow it to support UDP port.
`--tcp=[]: Port pairs can be specified as '<port>:<targetPort>'.`

Two possible ways to modify flag. I'd prefer the latter one.
- add a flag `--udp` which is similar to `--tcp`
- change flag `--tcp` to `--ports` which has format `<protocol>:<port>:<targetPort>`

Do I miss anything?

/cc @thockin @adohe 
/sig network
/sig cli
/assign @xiangpengzhao

### PR DESCRIPTION
**Pull request by guangxuli**
**What this PR does / why we need it**:
Just pick up pr #47485, this is helpful, we should pick it up since it has been close by k8s-robot, and borrow most of code. @sandoracs seems to inactive. :(

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

Fixes #47337 
**Special notes for your reviewer**:

**Release note**:

```release-note
Allow set UDP port when creating service
```
@xiangpengzhao

### CODE REVIEW

File: pkg/kubectl/cmd/create_service.go [hunk 1]
```diff
@@ -51,20 +51,22 @@ var (
...
-    kubectl create service clusterip my-cs --tcp=5678:8080
+    kubectl create service clusterip my-cs --ports=tcp:5678:8080
 
     # Create a new ClusterIP service named my-cs (in headless mode)
     kubectl create service clusterip my-cs --clusterip="None"`))
 )
 
 func addPortFlags(cmd *cobra.Command) {
 	cmd.Flags().StringSlice("tcp", []string{}, "Port pairs can be specified as '<port>:<targetPort>'.")
+	cmd.Flags().MarkDeprecated("tcp", "It will be removed in the future, please use --ports instead")
```
Comment (ericchiang) [inline pkg/kubectl/cmd/create_service.go]: Was this change discussed and signed of by sig-cli?

cc @kubernetes/sig-cli-pr-reviews

File: pkg/kubectl/cmd/create_service.go [hunk 2]
```diff
@@ -51,20 +51,22 @@ var (
 
 	serviceClusterIPExample = templates.Examples(i18n.T(`
     # Create a new ClusterIP service named my-cs
-    kubectl create service clusterip my-cs --tcp=5678:8080
+    kubectl create service clusterip my-cs --ports=tcp:5678:8080
```
Comment (ericchiang) [inline pkg/kubectl/cmd/create_service.go]: Why not add a `--udp` flag?

File: pkg/kubectl/cmd/create_service.go (hunk 1 above)
Comment (mengqiy) [reply pkg/kubectl/cmd/create_service.go]: I think this is fine.
@fabianofranz Thoughts?

File: pkg/kubectl/cmd/create_service.go (hunk 2 above)
Comment (guangxuli) [reply pkg/kubectl/cmd/create_service.go]: i think one generic flag is better than specific flag to expand the future function extension and also tcp/udp/ belong to the same class of tcp/ip protocol. If we support both --tcp and --udp flag we also need to think about the  mutual exclusion.

**Comment by guangxuli** [comment]
@kubernetes/sig-cli-pr-reviews

File: pkg/kubectl/cmd/create_service.go (hunk 2 above)
Comment (zjj2wry) [reply pkg/kubectl/cmd/create_service.go]: i think use `--protocol` and `--port` would be better

Comment (liggitt) [reply pkg/kubectl/cmd/create_service.go]: > i think use --protocol and --port would be better

that makes it difficult to expose multiple ports on different protocols

**Comment by guangxuli** [comment]
@kubernetes/sig-cli-pr-reviews

File: pkg/kubectl/cmd/create_service.go (hunk 2 above)
Comment (shiywang) [reply pkg/kubectl/cmd/create_service.go]: I vote for `--udp`

**Comment by guangxuli** [comment]
@kubernetes/sig-cli-pr-reviews
@fabianofranz do you have time slot to review this? or after the code freeze, hmm, seems ours guys pretty occupied these days. :smile:

**Comment by guangxuli** [comment]
@ericchiang @mengqiy should we have another round discussion for this? 
hi @fabianofranz  how do think about this? :)

**Comment by sandoracs** [comment]
Hi all,

Sorry for being so inactive lately. I am ready to rejoin to this task. :)

**Comment by guangxuli** [comment]
ping @derekchiang @mengqiy 
pending so long time, so we should take another look for this PR.

**Comment by derekchiang** [comment]
@guangxuli did you mean @ericchiang ?

File: pkg/kubectl/service_basic.go [hunk 3]
```diff
@@ -83,7 +83,9 @@ func (ServiceExternalNameGeneratorV1) ParamNames() []GeneratorParam {
 }
 
 func parsePorts(portString string) (int32, intstr.IntOrString, error) {
```
Comment (ericchiang) [inline pkg/kubectl/service_basic.go]: Should this return the protocol too? Avoid parsing logic later.

```diff
+		port, targetPort, err := parsePorts(portsString)
  		if err != nil {		  		if err != nil {
  			return nil, err		  			return nil, err
  		}		  		}
  		  
+		separatorPos := strings.Index(portsString, ":")
+		protocol := strings.ToUpper(portsString[0:separatorPos])
+		portName := strings.Replace(portsString[separatorPos+1:], ":", "-", -1)
```

File: pkg/kubectl/cmd/create.go [hunk 4]
```diff
@@ -285,6 +285,28 @@ func NameFromCommandArgs(cmd *cobra.Command, args []string) (string, error) {
 	return args[0], nil
 }
 
+// PortsFromCommandArgs is a function to gather ports config from args. It supports deprecating "--tcp" arg.
+func PortsFromCommandArgs(cmd *cobra.Command) ([]string, error) {
```
Comment (ericchiang) [inline pkg/kubectl/cmd/create.go]: I'd prefer we avoid exporting methods unless we explicitly need them to be exported.

File: pkg/kubectl/cmd/create.go [hunk 5]
```diff
@@ -285,6 +285,28 @@ func NameFromCommandArgs(cmd *cobra.Command, args []string) (string, error) {
...
+	var (
+		ports []string
+		err   error
+	)
+	if ports = cmdutil.GetFlagStringSlice(cmd, "ports"); len(ports) == 0 {
```
Comment (ericchiang) [inline pkg/kubectl/cmd/create.go]: nit, fewer nested conditions:

```go
ports := cmdutil.GetFlagStringSlice(cmd, "ports")
if len(ports) == 0 {
    ports = cmdutil.GetFlagStringSlice(cmd, "tcp")
    for i, port := range ports {
        ports[i] = "tcp:"+port
    }
}
if len(ports) == 0 {
    return nil, cmdutil.UsageErrorf(cmd, "at least one port is required")
}
return ports, nil
```

File: pkg/kubectl/cmd/create.go [hunk 6]
```diff
@@ -285,6 +285,28 @@ func NameFromCommandArgs(cmd *cobra.Command, args []string) (string, error) {
...
+		ports = cmdutil.GetFlagStringSlice(cmd, "tcp")
+		if len(ports) == 0 {
+			err = cmdutil.UsageErrorf(cmd, "PORTS or TCP is required")
```
Comment (ericchiang) [inline pkg/kubectl/cmd/create.go]: Don't mention the deprecated flag in the error message.

**Review by ericchiang — COMMENTED** [review]
Just a couple nits. Other than that works for me.

**Pull request closed without merging**